tests/lib/external/snapd-testing-tools/utils/spread-filter

```diff
@@ -0,0 +1,223 @@
+
+            # if the change is a file in the same dir than a task.yaml
+            change_dir = os.path.dirname(change_path)
+            if self._is_test_dir(change_dir):
+                return self._get_test_exec_dir(change_dir)
+            # if the change is a file which is not in the same dir than a task.yaml
+            else:
+                test_dir = self._get_parent_with_task(change_path)
+                return self._get_test_exec_dir(test_dir)
+        # if the change is deleted file/dir
+        else:
+            # if a task.yaml has been deleted
+            if self._is_task_file(change_path):
+                return ""
+            else:
+                test_dir = self._get_parent_with_task(change_path)
+                return test_dir
+
+    def _get_parent_with_task(self, change_path: str) -> str:
+        if self.verbose:
+            print("calculating parent with task {}".format(change_path))
+        parent_path = os.path.dirname(change_path)
+        while parent_path != "":
+            if os.path.isfile(os.path.join(parent_path, self.TASK)):
+                return parent_path
+            parent_path = os.path.dirname(parent_path)
+        return ""
+
+    def _clean_executions(self, executions: list[str]) -> list[str]:
+        if self.verbose:
+            print("cleaning executions for {}".format(executions))
+
+        final_executions = []
+        # Get the execution paths ordered starting from the longest one
+        sorted_executions = sorted(executions, key=len, reverse=True)
+        for i in range(len(sorted_executions)):
+            discard_execution = False
+            long_execution = sorted_executions[i]
+            if self.verbose:
+                print("  checking repeated executions for {}".format(long_execution))
+
+            for j in range(i + 1, len(sorted_executions)):
+                short_execution = sorted_executions[j]
+                if self.verbose:
+                    print(
+                        "    comparing repeated executions for {}".format(
+                            short_execution
+                        )
+                    )
+
+                # if the shorter executions is a test, we continue
+                if not short_execution.endswith("/"):
+                    if long_execution == short_execution:
+                        discard_execution = True
+                        break
+                # if the shorter execution is a either suite or group of suites
+                else:
+                    # if the long executions starts with short execution means
+                    # the long execution is included in the short one and it can
+                    # be discarded
+                    if long_execution.startswith(short_execution):
+                        discard_execution = True
+                        break
+
+            # take decision after all shorter executions have been compared
```

Comment:
Nitpick: could be more consistent with when we start comment with upper-case or not.

Reply:
Comments should start with upper-case, update almost all of them